classes/submission/PKPSubmissionDAO.inc.php

```diff
@@ -94,7 +94,9 @@ function _getCache() {
 	 */
 	function _fromRow($row) {
 		$submission = parent::_fromRow($row);
-		$submission->setData('publications', Services::get('publication')->getMany(['submissionIds' => $submission->getId()]));
+		$submission->setData('publications', iterator_to_array(
+			Services::get('publication')->getMany(['submissionIds' => $submission->getId()])
+		));
```

Comment:
Should objects attached to other objects be kept as `Iterator`s or changed to `array`s? For example a `publication` will have `authors` and `galleys` (and in OMP `publicationFormats` have lots of attached objects).

Will we save much memory by not instantiating them until needed?

Another performance question: what's the cost of instantiating and destroying an object? If I were to run the iterator twice:

```
$authors = $object->getData('iterator_property');
foreach ($authors as $author) {
  // do something
}
```

And then in another place somewhere else in the code I iterated again:

```
$authors = $object->getData('author');
foreach ($authors as $author) {
  // do something
}
```

Would it have been better for me to convert the authors to an array so that the objects were already instantiated the second time? Or is it generally better to iterate twice, even if the objects are being set up as they go?

Reply:
Ah, I was going to write up a note about this but forgot. These iterators can only be iterated once, from start to finish, no-rewindsies -- that's a requirement of the underlying database implementation. And they should always be iterated fully so that the query can be closed. So when storing fetched dependent data in a DataObject -- e.g. the author list, or the publication list -- yes, these should be converted to an array.

If you try to iterate twice -- e.g. a second foreach on the same data -- you'll get [an Exception](https://github.com/pkp/pkp-lib/pull/5123/files#diff-a9ca68ca5b8618d5c948b25946e5cf7aR69).

Reply:
Hmm, I worry that this is going to catch us and others by surprise. Even if we document this limitation, it's likely that what comes back from the service class will be conceptualised as a resource that can be used and re-used like other variables. Especially since we are trying to promote the service classes as an intuitive API for working with objects, it will be deployed by developers who are not yet familiar with our application structure.

I'd like to find a way to hold onto that simplicity if we can. As I understand it, we've got three kinds of use-cases here that we're trying to juggle. One is that we want to get a small collection of objects, typically based on their relationship to another object. For example: get all the authors/galleys for a publication.

The second is getting a limited set of objects from a large pool of objects. For example: get submissions for page 1 of the active submissions list.

The third is getting a complete set of a large pool of objects. For example: get all published submissions for this context.

For the first case, we usually convert to an array of objects and in the second case we may or may not want to do so depending on the potential number of objects. But the last case is definitely different and needs a different way to handle this.

What do you think about using a different service method, like:

```
$iterator = Services::get('submissions')->getIterator($args = []);
```

This would provide both options depending on what we want to return at the time.

We could go one step further and make wrapper functions which made the return explicit in the usage:

```
public function getArray($args = []) {
	return $this->_getResultFactory($args)->toArray();
}

public function getIterator($args = []) {
	return $this->_getResultFactory($args)->toIterator();
}

private function _getResultFactory($args = []) {
	$result = ...;
	return new DAOResultFactory($result, $dao, '_fromRow');
}
```

The downside here is that it's probably easier for new developers to use `getArray()` when they should probably use `getIterator()`. But the upside is that we keep a pretty legible API that obscures some of the complexity of the result factory, so that it is easier to pick up and work with.

Thoughts?

Reply:
> Even if we document this limitation, it's likely that what comes back from the service class will be conceptualised as a resource that can be used and re-used like other variables.

Just to illustrate this: even after all these years I didn't know that a `DAOResultFactory` could not be cycled twice... :joy:

Reply:
The codebase made heavy use of direct returns of `DAOResultFactory` objects until the service classes established a new pattern of arrays, so the pattern is proven in terms of meeting calling code's requirements -- 99% of the time we iterate through the list once and that's it. Is it more an issue of coder expectations based on the Iterator class?

If so, maybe a better fit would be a Generator, since those are intended to be one-time consumable and forward-only. From https://www.php.net/manual/en/language.generators.comparison.php:
>  generators are forward-only iterators, and cannot be rewound once iteration has started. This also means that the same generator can't be iterated over multiple times: the generator will need to be rebuilt by calling the generator function again.